### PR TITLE
Preliminary support for modules/externals + CLI subcommands

### PR DESCRIPTION
**Pull request by AltGr**
- Module support is far from ready yet, but since it doesn't seem to break the
  existing in its current state, it may be worth merging already to avoid
  conflicts later on

- There are a few refactors that may also be worthwile (rename of `IdentName` to
  `Ident`, more uses of the AST mapper)

- Included, and maybe most important here is a refactor of command-line parsing
  to use sub-commands ; this allows different backends but also plugins to have
  specific flags. See 52b1c25d5 for details.

I can take the time to split the 3rd point into a separate PR if more discussion
is needed for either that or the rest, or for easier review if having all this
bundled is a problem.

### CODE REVIEW

File: compiler/shared_ast/interpreter.ml [hunk 1]
```diff
@@ -375,10 +390,226 @@ let rec evaluate_operator
...
+        Obj.repr (fun x ->
+            curry (runtime_to_val eval_expr ctx m targ x :: acc) targs)
+    in
+    curry [] targs
+  | _ -> assert false
+
+(* let f e = (e : (< .. > as 'a, 't) gexpr :> (< custom : _; 'a; .. >, 't) gexpr ) 
+ * 
+ * let f (type a) ((e: (< custom: a; .. >, 't) naked_gexpr), t) =
+ *   if false then ECustom { obj = Obj.repr (); targs = []; tret = (TLit TUnit, Pos.no_pos) }, t
```
Comment (denismerigoux) [inline compiler/shared_ast/interpreter.ml:510]: should this dead code be removed?

File: compiler/shared_ast/interpreter.ml [hunk 2]
```diff
@@ -552,6 +804,142 @@ let rec evaluate_expr :
...
+ *       ; scopeVarStates : 'g
+ *       ; scopeVarSimpl : 'h
+ *       ; explicitScopes : 'i
+ *       ; assertions : 'j
+ *       ; defaultTerms : 'k
+ *       ; exceptions : 'l
+ *       ; custom : 'm }
+ *       ->
+ *       (< monomorphic : 'a
+ *        ; polymorphic : 'b
```
Comment (denismerigoux) [inline compiler/shared_ast/interpreter.ml:854]: dead code also?

File: compiler/shared_ast/interpreter.ml [hunk 3]
```diff
@@ -552,6 +804,142 @@ let rec evaluate_expr :
...
+    | ( ( EAssert _ | ELit _ | EApp _ | EArray _ | EVar _ | EExternal _ | EAbs _
+        | EIfThenElse _ | ETuple _ | ETupleAccess _ | EInj _ | EStruct _
+        | EStructAccess _ | EMatch _ ),
+        _ ) as e ->
+      Expr.map ~f e
+    | _ -> .
+  in
+  Expr.unbox (f e)
+
+let evaluate_expr ctx e = delcustom (evaluate_expr ctx (addcustom e))
```
Comment (denismerigoux) [inline compiler/shared_ast/interpreter.ml:941]: this shenanigan could gain from some explanations :)

File: compiler/shared_ast/optimizations.ml [hunk 4]
```diff
@@ -184,7 +184,7 @@ let rec optimize_expr :
       when name = name1 ->
       Mark.remove (StructField.Map.find field fields)
     | EDefault { excepts; just; cons } -> (
-      (* TODO: mechanically prove each of these optimizations correct :) *)
+      (* TODO: mechanically prove each of these optimizations correct *)
```
Comment (denismerigoux) [inline compiler/shared_ast/optimizations.ml:187]: :(

File: compiler/shared_ast/qident.mli [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+(** This module defines module names and path accesses, used to refer to
+    separate compilation units. *)
+
+type modname = string
+(** Expected to be a uident *)
```
Comment (denismerigoux) [inline compiler/shared_ast/qident.mli:21]: That means you'll change these afterwards?

**Review by denismerigoux — APPROVED** [review]
Thanks, I reviewed it all and it's fine to merge bundled like this!

File: compiler/shared_ast/qident.mli (hunk 5 above)
Comment (AltGr) [reply compiler/shared_ast/qident.mli:21]: No, just that the string is assumed to start with an uppercase letter

File: compiler/shared_ast/optimizations.ml (hunk 4 above)
Comment (AltGr) [reply compiler/shared_ast/optimizations.ml:187]: (my editor was'nt clever enough to pair the parens of the comment regardless of the emoji… now I feel bad for killing an innocent smile :cry: )

File: compiler/desugared/from_surface.ml [hunk 6]
```diff
@@ -1248,37 +1248,55 @@ let process_topdef
...
   let translate_tbase (tbase, m) = translate_typ (Base tbase, m) in
   let typ = translate_typ def.S.topdef_type in
-  let expr =
-    match def.S.topdef_args with
-    | None -> translate_expr None None ctxt def.S.topdef_expr
-    | Some (args, _) ->
+  let expr_opt =
+    match def.S.topdef_expr, def.S.topdef_args with
+    | None, _ -> None
+    | Some e, None -> Some (Expr.unbox (translate_expr None None ctxt e))
```
[the remark this answers is not in the document]
Comment (AltGr) [reply compiler/desugared/from_surface.ml:1259]: Here we are dealing with top-level definitions (before program linearisation), so they are expected to be closed.

**Comment by AltGr** [comment]
Ouch indeed I had left some dead bodies lying around :zombie: 
Now cleaned up and added some clarifications (plus a function `Expr.unbox_closed` that adds an assertion to `Expr.unbox`)

**Pull request merged**